dvc/remote/base.py

```diff
@@ -257,8 +239,25 @@ def _download_dir(
                 )
             )
             with ThreadPoolExecutor(max_workers=self.remote.JOBS) as executor:
-                futures = executor.map(download_files, from_infos, to_infos)
-                return sum(futures)
+                futures = [
+                    executor.submit(download_files, from_info, to_info)
+                    for from_info, to_info in zip(from_infos, to_infos)
+                ]
+
+                # NOTE: unlike pulling/fetching cache, where we need to
+                # download everything we can, not raising an error here might
+                # turn very ugly, as the user might think that he has
+                # downloaded a complete directory, while having a partial one,
+                # which might cause unexpected results in his pipeline.
+                for future in as_completed(futures):
+                    # NOTE: executor won't let us raise until all futures that
+                    # it has are finished, so we need to cancel them ourselves
+                    # before re-raising.
+                    exc = future.exception()
+                    if exc:
+                        for entry in futures:
+                            entry.cancel()
+                        raise exc
 
```

Comment:
Also, just keeping this consistent for now. In the future we might want to condier adding something like `onerror` in `os.walk`, which will let us know if we should somehow process the exception and continue or abort.

Reply:
Though, probably better to try to download to a temporary location first and delete it if unsuccessful. That would be consistent for now. Will look into Output.download() too.